cmd/snap/cmd_run.go

```diff
@@ -489,22 +489,36 @@ func (x *cmdRun) straceOpts() (opts []string, raw bool, err error) {
 	return opts, raw, nil
 }
 
-// isSnapRefreshConflictDetected detects if snap refreshed was started while not
-// holding the inhibition hint file lock.
+// checkSnapRunInhibitionConflict detects if snap refreshed/removed was started
+// while not holding the inhibition hint file lock.
 //
 // For context, on snap first install, the inhibition hint lock file is not created
-// so we cannot hold it. It is created after the first refresh. This allows for a
-// window where we don't hold the lock before the tracking cgroup is created where
-// a snap refresh could start.
-func isSnapRefreshConflictDetected(app *snap.AppInfo, hintFlock *osutil.FileLock) bool {
-	if !features.RefreshAppAwareness.IsEnabled() || app.IsService() || hintFlock != nil {
+// so we cannot hold it. It is created after the first refresh/remove. This allows
+// for a window where we don't hold the lock before the tracking cgroup is created
+// where a snap refresh/removal could start.
+func checkSnapRunInhibitionConflict(app *snap.AppInfo) error {
+	// Remove hint check takes precedence because we want to exit
+	snapName := app.Snap.InstanceName()
+	hint, _, err := runinhibit.IsLocked(snapName)
+	if err != nil {
+		return err
+	}
```

Comment:
hintFlock when `checkSnapRunInhibitionConflict` is reached is always nil, check `closeFlockOrCheckConflict`.
Reaching `checkSnapRunInhibitionConflict` indicates that we have a corner case where the snap was installed and didn't go through refreshes, so no inhibition hint file existed for it which is why we have hintFlock as nil. That's why we need to do the race condition checks, in case of remove inhibition just exit and for anything else retry.